calico/felix/endpoint.py

```diff
@@ -81,6 +102,86 @@ def _create(self, combined_id):
+        else:
+            del self.tier_orders[tier]
+
+        new_tier_sequence = sorted(self.tier_orders.iterkeys(),
+                                   key=lambda k: (self.tier_orders[k], k))
+        if self.tier_sequence != new_tier_sequence:
+            _log.info("Sequence of profile tiers changed, refreshing all "
+                      "endpoints")
+            self.tier_sequence = new_tier_sequence
+            self.endpoints_with_dirty_policy.update(
+                self.endpoints_by_id.keys()
+            )
+            self._update_dirty_policy()
+
+    @actor_message()
+    def on_prof_labels_set(self, profile_id, labels):
+        _log.debug("Profile labels updated for %s: %s", profile_id, labels)
+        self._label_inherit_idx.on_parent_labels_update(profile_id, labels)
+        self._update_dirty_policy()
+
+    @actor_message()
+    def on_policy_selector_update(self, policy_id, selector_or_none,
+                                  order_or_none):
+        _log.debug("Policy %s selector updated to %s (%s)", policy_id,
+                   selector_or_none, order_or_none)
+        if selector_or_none != self.selectors_by_pol_id.get(policy_id):
+            if selector_or_none is not None:
+                self.selectors_by_pol_id[policy_id] = selector_or_none
+            else:
+                del self.selectors_by_pol_id[policy_id]
+            # Defer to the label index, which will call us back via
+            # on_policy_match_started and on_policy_match_stopped.
+            self.policy_index.on_expression_update(policy_id,
+                                                   selector_or_none)
+
+        # Before we update the policies, check if the order has changed,
+        # which would mean we need to refresh all endpoints with this policy
+        # too.
+        if order_or_none != self.profile_orders.get(policy_id):
+            if order_or_none is not None:
+                self.profile_orders[policy_id] = order_or_none
+            else:
+                del self.profile_orders[policy_id]
+            self.endpoints_with_dirty_policy.update(
+                self.policy_index.matches_by_expr_id.iter_values(policy_id)
+            )
+
+        # Finally, flush any updates to our waiting endpoints.
+        self._update_dirty_policy()
+
+    def on_policy_match_started(self, expr_id, item_id):
+        _log.info("Policy %s now applies to endpoint %s", expr_id, item_id)
+        self.pol_ids_by_ep_id.add(item_id, expr_id)
+        self.endpoints_with_dirty_policy.add(item_id)
+
+    def on_policy_match_stopped(self, expr_id, item_id):
+        _log.info("Policy %s no longer applies to endpoint %s",
+                  expr_id, item_id)
+        self.pol_ids_by_ep_id.discard(item_id, expr_id)
+        self.endpoints_with_dirty_policy.add(item_id)
+
     def _on_object_started(self, endpoint_id, obj):
         """
         Callback from a LocalEndpoint to report that it has started.
```

Comment:
That's part of the life-cycle of a ref-counted actor.  Once it's initialised on its own thread it calls back to the manager to say it's ready.